Add set:cal service with a few cal functions.

ITotalJustice opened the pull request:
Hi, using [switchbrew docs](https://switchbrew.org/wiki/Settings_services#set:cal) I added the set:cal service and a few of the functions that were documented.

I have tested them and they work.

Let me know if you would like me to rename anything or move the functions around.

yellows8 commented:
Added the structs for these to wiki, use those instead of buffer/size params.

fincs commented on nx/include/switch/services/set.h:
```diff
@@ -271,6 +271,36 @@ void setExit(void);
 Service* setGetServiceSession(void);
 
+/// Initialize setcal.
+Result setcalInitialize(void);
```
Put the new functions at the end of the file (same thing for the .c)

ITotalJustice commented:
Thanks for the feedback! I adding the structs, let me know if they're okay.

I noticed that all but the gamecard cert first `0x4` seems to be the offset of the actual data from after `0x4`. So data offset is `0x4 + offset`.

Also, the certs have some data in the same offsets / same values. For example the `SetCalEccB233DeviceCertificate` has the deviceID at offset `0xC4` which is also the same as the `SetCalRsa2048DeviceCertificate`.

[I have a basic example of the deviceID struct](https://github.com/ITotalJustice/NX-DeviceID-Viewer/blob/588f79f01de4840b73287c52a3e9777bdb8e5e1a/source/main.c#L15) (and also the `SetCalEccB233DeviceCertificate`). I could add this if you'd like, but just let me know how you'd like me to name it.

ITotalJustice commented:
I so far found two types of `NXCA`, I guess these should be an enum for the different versions?

`NintendoNXCA1Prod1` found in the eci_device_cert and ecc_b233_device_cert.
`NintendoNXCA2Prod10` found in the ssl_cert.

And lastly, looking at the gamecard cert, in the docs is says it has a cal container struct, however the first `0x4`do not appear to be a size (or an offset, like with the rest of the certs) as the number is much too large, its `0x721BE113` for me.

hexkyz commented:
@ITotalJustice I've updated the switchbrew wiki (https://switchbrew.org/wiki/Settings_services#set:cal) with all the sizes and names for these structs. Note that the "container" structure is only used outside of CAL0. The certificate data itself is opaque and copied directly from it's offset in calibration data, but the settings system module encapsulates this data by adding 4 extra bytes to describe it's size.

ITotalJustice commented:
@hexkyz Thank you for adding them, it's very helpful. I have added the structs using your information, I hope it looks okay, I will change them if not.

Yes i had noticed that sys adds `0x4` to the start of the keys, which is the size of the entire key including the '0x4' added. I have added this to the structs (that i have tested) to reflect this.

I still don't quite understand the GameCardCert. In the docs, it says the first `0x4` is the size of the cert, but the value at 0x4 does not represent this.

yellows8 commented on nx/include/switch/services/set.h:
```diff
@@ -262,6 +262,128 @@ typedef struct {
+} SetCalRsa2048DeviceCertificate;
+
+typedef struct {
+    size_t size;            ///< Size of the entire key.
```
If this is 32bit use u32, same below (otherwise use u64).

ITotalJustice replied:
oops! Good catch.

ITotalJustice replied:
I have fixed this.

yellows8 commented on nx/include/switch/services/set.h:
```diff
@@ -885,3 +1007,33 @@ Result setsysGetTouchScreenMode(SetSysTouchScreenMode *out);
+/// Gets the Service object for the actual setcal service session.
+Service* setcalGetServiceSession(void);
+
+/// Gets EciDeviceCertificate of size 0x180.
```
`Gets the \ref {full struct name}.`

ITotalJustice replied:
I have updated the comments to match that format.

yellows8 commented on nx/include/switch/services/set.h:
```diff
@@ -262,6 +262,128 @@ typedef struct {
+} SetCalBdAddress;
+
+typedef struct {
+    u8 cfgId1[0x1E];
```
`u8 cfg[...]`

yellows8 commented on nx/include/switch/services/set.h:
```diff
@@ -262,6 +262,128 @@ typedef struct {
+} SetCalCountryCode;
+
+typedef struct {
+    u32 offset;                     ///< Relative to current position.
```
Align these comments with each other (similar to elsewhere).

ITotalJustice replied:
I have aligned them, hope it looks okay.

hexkyz commented:
@ITotalJustice Oh, you mean this? https://switchbrew.org/wiki/Settings_services#GameCardCertificate
I see the confusion, the maximum size of the gamecard data is 0x400 and this is how much is read from CAL0. So it shouldn't be "it's maximum size is 0x400 bytes" but "it's maximum size is 0x404 bytes".
